Buffered multi-metric packets

vmg opened the pull request:
This should greatly increase the throughput of the StatsD client and server. See the detailed commit messages for explanations and design decisions.

cc @ross @aroben

aroben commented on lib/statsd.rb:
```diff
@@ -140,7 +201,7 @@ def send(stat, delta, type, sample_rate=1)
       stat.gsub!(RESERVED_CHARS_REGEX, "_".freeze)
 
-      msg = ""
+      msg = String.new
```
What's the reason for this change? I didn't see anything in the commit messages. (My guess it's to be compatible with frozen-by-default string literals in a future Ruby version.)

vmg replied:
Correct guess! In 2.3 (if you enable frozen by default), and in 3.0 (for all strings), the empty string will come as frozen, so we won't be able to append to it. This small change will hopefully save us a small version bump! :)

(in the future we should also be able to remove all the `.freeze` calls in the file, but they leaving them there causes no harm at all)

aroben commented on lib/statsd.rb:
```diff
@@ -14,14 +14,77 @@
+    def flush!
+      return unless @buffer.bytesize > 0
+      sock.write(@buffer)
+      @buffer.clear
```
It [looks like `String#clear` deallocates the buffer](https://github.com/ruby/ruby/blob/9efd4dbe73babbc1be41d4bf72db2dbbcb1cfa5d/string.c#L1841-L1850). Is there a way we can keep the buffer around?

vmg replied:
:angel: Based @aroben :angel: 

This is indeed very counter intuitive behavior that will cause needless GC churn. Let me investigate if there's an alternative that will set the string size to 0.

cc @charliesome

vmg replied:
Hmpf. No luck here. Our only options are writing a C extension or eating the needles GC churn. Thoughts?

vmg commented:
@aroben: Pretty happy with the refactoring in the first commit. It seems like we added "Secure Mode" _before_ we added support for custom Client Classes, so the logic for the secure mode was left mushed up inside the main class. I think keeping all the Statsd-secure code inside a custom client class makes it look much cleaner!

aroben commented on lib/statsd.rb:
```diff
@@ -14,14 +14,77 @@
+      sock.write(@buffer)
+      @buffer.clear
+    rescue => boom
+      nil
```
Is there a more restricted set of exceptions we can rescue here?

vmg replied:
This `rescue` has been cargo-culted all the way from the original gem. I agree we should narrow this down to UDPSocket exceptions. Let me investigate if there's a definite list of those.

aroben commented on lib/statsd.rb:
```diff
@@ -14,14 +14,77 @@
+      flush! if @buffer.bytesize + msg.bytesize > BufferedUDPClient.max_packet_size
+      @buffer << msg
+      @buffer << "\n".freeze
+      nil
```
If `@buffer.bytesize + msg.bytesize == BufferedUDPClient.max_packet_size`, we won't `flush!` on this `send`. But because of the newline, the next `send` call will actually send `BufferedUDPClient.max_packet_size + 1` bytes.

Also, what happens if `@buffer.bytesize >= BufferedUDPClient.max_packet_size`? Should we prevent that somehow?

aroben replied:
> If `@buffer.bytesize + msg.bytesize == BufferedUDPClient.max_packet_size`, we won't `flush!` on this `send`. But because of the newline, the next `send` call will actually send `BufferedUDPClient.max_packet_size + 1` bytes.

Oh, I see now that this is intended because `BufferedUDPClient.max_packet_size` is 511, not 512. That seems a bit misleading; the maximum packet we want to send is 512 bytes, not 511, so it's weird for the constant to be 511.

vmg replied:
> But because of the newline, the next send call will actually send BufferedUDPClient.max_packet_size + 1 bytes.

This is safe -- the default `max_packet_size` has plenty of headroom to fit inside the most strict MTUs.

> Also, what happens if @buffer.bytesize >= BufferedUDPClient.max_packet_size? Should we prevent that somehow?

I think you mean what happens if `msg.bytesize` doesn't fit in a packet? The case you've written should never happen. If `msg.bytesize >= max_packet_size`, we will transmit a packet which may or may not be too large... There's nothing we can do besides sending the packet and praying that the network doesn't drop it (since `max_packet_size` is not a hard cap). I think always sending the packet (even if it's potentially too large) is a better option than silently discarding it. I'd rather give the network a chance to drop it!

aroben replied:
And since the max is configurable, it seems especially important for it to be the actual max, not the `max - 1`.

aroben replied:
Sending a too-large packet seems like an OK behavior. We could consider sending it immediately rather than waiting for the next `send` call, since we know we're guaranteed to `flush!` on the next `send` anyway.

vmg replied:
Right. I agree. In practice this won't matter, because this is based around MTUs and affected by the network's congestion, but I agree that the semantics should be right. Let me fix it.

ross commented on lib/statsd.rb:
```diff
@@ -14,19 +14,82 @@
-  class RubyUdpClient
-    attr_reader :key, :sock
+  class BufferedUDPClient < UDPClient
+    @max_packet_size = 512
```
Piling on late here, but if this is really just the max size that will be batched perhaps the name is a little confusing. Maybe max_batch_size rather than packet? Since we could send out a packet that's larger if someone sends us a huge key or something (by accident.)

aroben commented on lib/statsd.rb:
```diff
@@ -128,8 +165,24 @@ def time(stat, sample_rate=1)
+    return unless @buffering
+    @shards.map! { |client| client.base_client }
+    @buffering = false
+  end
```
Maybe this should also call `flush_all!`? I suppose that could be left to the caller; perhaps a comment explaining that would be good.

vmg replied:
Flushing now!

jnunemaker commented on lib/statsd.rb:
```diff
@@ -165,28 +220,28 @@ def select_shard(stat)
-    signature + payload
-  end
+  class Buffer
+    DEFAULT_BUFFER_CAP = 512
```
Any reason for choosing this? Just curious if it is intentional or a best guess.

vmg replied:
@jnunemaker: Although y'all in dotcom usually lean towards long PR bodies, in Systems (given that we contribute a lot of code upstream) we go for small, well defined commits and detailed commit messages: https://github.com/github/statsd-ruby/commit/a06bb3ad2128ef03711a38639ae338cab03afc4b ;)

vmg replied:
...However, if you think it'd be easier for other people in the future, I can copy paste the commit message as a comment.

jnunemaker replied:
Duh. Forgot about that. Should have read the commits. Thanks!
